### ISSUE TITLE
Clarification for discrete inputs in an FMU

### ISSUE THREAD

**Issue by modelica-trac-importer** [modelica/fmi-standard#123]
**Reported by awittkop on 14 Jan 2013 18:47 UTC**
Can a change in a discrete input trigger an event in an FMU?
It would seem to be that the answer is no, as we have:
pg. 20: Discrete values can only change at an event instance
pg. 40: Discrete variables are constant between internal events

There is a discussion around pg. 64 that seems to indicate that if a discrete input changes, then the FMU needs to wait for the next internal event to apply the change.
Is this the case? If so, it should be stated explicitly.

Any clarification would be welcome.

----

Migrated-From: https://trac.fmi-standard.org/ticket/123

**Comment by modelica-trac-importer** [comment]
**Comment by anonymous on 24 Jan 2013 09:54 UTC**
In FMI for Model Exchange, the environment has to enter event handling status as soon as a discrete value, which an input of a FMU changes its value. Please have a look at the Figure 6 Calling sequence of Model Exchange C functions. The sequence is:
- the environment calls fmiSetXXX for all discrete inputs, that have change (the state "set inputs" is entered), 
- fmiEventUpdate is called afterwards, which enters the state event pending.
In the state chart, this is called an "input event".
In other words, the FMU does not have to wait until the next internal event happens to apply the change of discrete inputs. The environment has to do event handling for the FMU when discrete inputs change.

We have to describe that behavior more clearly in the specification text.

**Comment by modelica-trac-importer** [comment]
**Comment by torstenblochwitz on 7 Feb 2013 14:57 UTC**
Torsten will make proposal for spec.

**Comment by modelica-trac-importer** [comment]
**Comment by otter on 11 Oct 2013 11:28 UTC**
Fixed in FMI 2.0: Explicitly introduced "external events" (so events triggered by the environment, e.g., because an input changes discontinuously). In such a case the environment has to stop the integration, has to call "fmiEnterEventMode" and then call fmiSetXXX on the input (providing the new, discontinuous, value of the input). So the model is called at the end of the integration with the "left limit" of the input, and with fmiSetXXX in fmiEnterEventMode with the "right limit".
